secretsauce/apps/portal/views.py

```diff
@@ -91,113 +87,74 @@ class ProjectDetail(generics.RetrieveUpdateDestroyAPIView):
+            except KeyError:
+                raise ParseError(detail="missing parameter: data_block")
+            except DataBlock.DoesNotExist:
+                raise Http404
+            self.perform_create(serializer)
+            constraint_block = get_object_or_404(ConstraintBlock.objects.all(), id=serializer.data['id'])
+            constraint_params = [ConstraintParameter(item_id=header.item_id, constraint_block=constraint_block) for header in data_block.schema.all()]
+            ConstraintParameter.objects.bulk_create(constraint_params)
+            serializer = self.get_serializer(instance=self.get_queryset().get(id=serializer.data['id']))
+
+            project = constraint_block.project
+            if project.cost_sheet:
+                items = project.items.all()
+                for constraint_param in serializer.data['params']:
+                    try:
+                        item_id = constraint_param.get('item_id')
+                        item_name = items.get(item_id=item_id).name
+                    except Item.DoesNotExist:
+                        item_name = None
+                    constraint_param['item_name'] = item_name
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
 
-class ConstraintParameterDetail(generics.RetrieveUpdateDestroyAPIView):
+class ConstraintBlockItems(generics.ListAPIView):
     """
-    Retrieve/Update/Destroy a constraint parameter
+    Retrieve ConstraintParameters associated to a ConstraintBlock
     """
     permission_classes = [IsOwnerOrAdmin]
     queryset = ConstraintParameter.objects.all()
     serializer_class = ConstraintParameterSerializer
 
-class ConstraintRelationshipCreate(generics.CreateAPIView):
-    """
-    Create ConstraintParameterRelationship
-    """
-    queryset = ConstraintParameterRelationship.objects.all()
-    serializer_class = ConstraintParameterRelationshipSerializer
+    def list(self, request, *args, **kwargs):
+        queryset = self.filter_queryset(self.get_queryset())
+
+        page = self.paginate_queryset(queryset)
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
+
+        serializer = self.get_serializer(queryset, many=True)
+        project = ConstraintBlock.objects.get(id=self.kwargs.get('pk')).project
+        if project.cost_sheet:
+            items = project.items.all()
+            for constraint_param in serializer.data:
+                try:
+                    item_id = constraint_param.get('item_id')
+                    item_name = items.get(item_id=item_id).name
+                except Item.DoesNotExist:
+                    item_name = None
+                constraint_param['item_name'] = item_name
+        return Response(serializer.data)
```

Comment:
Can explain what this section is for?

Reply:
Item names are obtained from FullItemRec.csv (aka cost sheet). There is only 1 cost sheet uploaded per project. This is to find the item names from the project items and match according to item_id